client/scala/src/main/scala/ai/verta/repository/Commit.scala

```diff
@@ -14,15 +17,113 @@ class Commit(
   private val commit: VersioningCommit, private val commitBranch: Option[String] = None
 ) {
   private var saved = true // whether the commit instance is saved to database, or is currently being modified.
+  private var loadedFromRemote = false // whether blobs has been retrieved from remote
+  private var blobs = Map[String, VersioningBlob]() // mutable map for storing blobs
 
   /** Return the id of the commit */
-  def id = commit.commit_sha.get
+  def id = commit.commit_sha
 
   override def equals(other: Any) = other match {
-    case other: Commit => commit.commit_sha.get == other.commit.commit_sha.get
+    case other: Commit => id.isDefined && other.id.isDefined && id.get == other.id.get
     case _ => false
   }
 
+  /** Retrieves the blob at path from this commit
+   *  @param path location of a blob
+   *  @return The blob. If not existed, or retrieving blobs from backend fails, return a failure.
+   */
+  def get(path: String)(implicit ec: ExecutionContext): Try[Blob] = {
+    loadBlobs().flatMap(_ =>
+      blobs.get(path) match {
+        case None => Failure(new NoSuchElementException("No blob was stored at this path."))
+        case Some(blob) => Success(versioningBlobToBlob(blob))
+      }
+    )
+  }
+
+  /** Adds blob to this commit at path
+   *  If path is already in this Commit, it will be updated to the new blob
+   *  @param path Location to add blob to
+   *  @param blob Instance of Blob subclass.
+   *  @return The new commit, if succeeds.
+   */
+  def update(path: String, blob: Blob)(implicit ec: ExecutionContext): Try[Commit] = {
+    loadBlobs().map(_ => {
+      // creating new commit:
+      val childCommit = getChild()
+
+      /** TODO: Add blob subtypes to pattern matching */
+      val versioningBlob = blob match {
+        case pathBlob: PathBlob => PathBlob.toVersioningBlob(pathBlob)
+      }
+
+      childCommit.blobs = blobs + (path -> versioningBlob)
+      childCommit.saved = false
+      childCommit.loadedFromRemote = true
+
+      childCommit
+    })
+  }
+
+  /** Retrieve commit's blobs from remote
+   */
+  private def loadBlobs()(implicit ec: ExecutionContext): Try[Unit] = {
+    if (!loadedFromRemote) {
+      // if the commit is not saved, get the blobs of its parent(s)
+      val ids: List[String] = commit.commit_sha match {
+        case Some(v) => List(v)
+        case None => commit.parent_shas.get
+      }
+
+      Try(ids.map(id => loadBlobsFromId(id)).map(_.get).reduce(_ ++ _)) match {
+        case Failure(e) => Failure(e)
+        case Success(map) => Success {
+          loadedFromRemote = true
+          blobs = map
+        }
```

Comment:
This information is being lost since you're converting to Unit. Is this information not used anywhere? I find it suspicious that the result of `Success(map)` is not used anywhere and the tests pass. Either something is mutable or the tests are wrong. Even the places that call `loadBlobs` ignore the return values.

Reply:
Hmm I think I mentioned it somewhere, but this is to simulate the lazy loading behavior of blobs. loadedFromRemote indicate whther the blobs has been loaded. So here, if the "map" is successfully loaded, it will be set to true, and the map itself assign to blobs.

Reply:
I'm aware that scala has "lazy val", which is basically this. But using that means I cannot reassign the variable blobs to another value, which makes initializing a new Commit instance difficult.

Reply:
This is also very similar to the initializing of S3/PathBlob btw. The extra complexity comes from the fact that we're trying to have the blobs map lazy. 

Otherwise I can use a very similar pattern with an apply method in the companion object: retrieve the blobs from DB, turn it into a map, then assign it to the blobs var of a the new Commit instance.

Reply:
To demonstrate my idea:

```
  def apply(
    clientSet: ClientSet,
    repo: Repository,
    versioningCommit: VersioningCommit,
    commitBranch: Option[String] = None,
    blobs = Option[Map[String, VersioningBlob]] = None
  ) = {
    var commit = new Commit(clientSet, repo, versioningCommit, commitBranch)
    commit.blobs = blobs.getOrElse(commit.loadBlobs())
    commit
  }
```
Something like this? I haven't tried actually using this, but the general idea would be this (assuming we sacrifice the laziness of commit.blobs)

Reply:
It's fine for now. But if I had to ask and you have to use 3 responses to clarify, then it deserves at least a few comments in the code explaining what's going on. We should also not return a struct that is discarded as that's very confusing.